File: epi_judge_java/epi/RealSquareRoot.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

public class RealSquareRoot {

    @EpiTest(testDataFile = "real_square_root.tsv")
    public static double squareRoot(double x) {
        double left = -1;
        double right = x + 1;

        while (right - left > 1e-8) {
            double mid = (left + right) / 2;
            if (mid * mid > x) {
                right = mid;
            } else {
                left = mid;
            }
        }

        return left;
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "RealSquareRoot.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/real_square_root.cc
#include "test_framework/generic_test.h"

double SquareRoot(double x) {
    if (x < 0) return -1;
    if (x == 0) return x;
    long double left = 0.0;
    long double right = 1e18;
    while (right - left > 1e-8) {
        long double candidate = (left + right) / 2;
        if (candidate * candidate > x) {
            right = candidate;
        } else {
            left = candidate;
        }
    }

    return left;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"x"};
    return GenericTestMain(args, "real_square_root.cc", "real_square_root.tsv",
                           &SquareRoot, DefaultComparator{}, param_names);
}
